scipy/stats/stats.py

```diff
@@ -4229,50 +4230,128 @@ def mannwhitneyu(x, y, use_continuity=True):
     use_continuity : bool, optional
             Whether a continuity correction (1/2.) should be taken into
             account. Default is True.
+    use_exact : bool, optional
+            Whether an exact test should be performed on the data.
+            See notes for default behavior.
 
     Returns
     -------
-    statistic : float
-        The Mann-Whitney statistics.
-    pvalue : float
-        One-sided p-value assuming a asymptotic normal distribution.
+    u : float
+        The Mann-Whitney U statistic.
+    p : float
+        p-value.
 
     Notes
     -----
-    Use only when the number of observation in each sample is > 20 and
-    you have 2 independent samples of ranks. Mann-Whitney U is
-    significant if the u-obtained is LESS THAN or equal to the critical
-    value of U.
+    Exact tests should be used for smaller sample sizes. Concretely, as
+    len(x) and len(y) increase to and beyond 8, the distribution of U differs
+    negligibly from the normal distribution[1]. The default behavior of this
+    test is to calculate the number of possible sequences for inputs of
+    length(x) and length(y), and to do an exact calculation if the number of
+    possible combinations is <100000. The default behavior may be overridden
+    with the use_exact flag.
+
+    This test corrects for ties and by default uses a continuity correction
+    when approximating the U statistic distribution.
+
+    The reported p-value is for a two-sided hypothesis. To get the one-sided
+    p-value set alternative to 'greater' or 'less' (default is 'two-sided').
 
-    This test corrects for ties and by default uses a continuity correction.
-    The reported p-value is for a one-sided hypothesis, to get the two-sided
-    p-value multiply the returned p-value by 2.
+    The reported U statistic mirrors the behavior of the wilcoxon.test
+    procedure of the R stats package v3.2.0, always reporting min(u1, u2),
+    even in one-sided tests.
 
+    .. versionadded:: 0.17.0
+
+    References
+    ----------
+    .. [1] HB Mann and DR Whitney (1947). "On a Test of Whether one of Two
+            Random Variables is Stochastically Larger than the Other".
+            Annals of Mathematical Statistics. doi:10.1214/aoms/1177730491
     """
     x = asarray(x)
     y = asarray(y)
     n1 = len(x)
     n2 = len(y)
-    ranked = rankdata(np.concatenate((x, y)))
-    rankx = ranked[0:n1]  # get the x-ranks
-    u1 = n1*n2 + (n1*(n1+1))/2.0 - np.sum(rankx, axis=0)  # calc U for x
-    u2 = n1*n2 - u1  # remainder is U for y
-    bigu = max(u1, u2)
-    smallu = min(u1, u2)
+    Stats = namedtuple('Stats', ['u', 'p'])
+    if use_exact == 'auto':
+        use_exact = (n1 < 10 or n2 < 10) and n1 + n2 < 100000 \
```

Comment:
This line limits the factorial calculation to a worst-case scenario of `100000! / (99999! * 1!) `, which can take up to a few seconds.